presto-main/src/main/java/com/facebook/presto/server/remotetask/WorkerTaskStatusFetcher.java

```diff
@@ -0,0 +1,284 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+class WorkerTaskStatusFetcher
+        implements SimpleHttpResponseCallback<List<TaskStatus>>
+{
+    private final ConcurrentHashMap<TaskId, ContinuousBatchTaskStatusFetcher.Task> idTaskMap;
+
+    private static final Logger log = Logger.get(ContinuousTaskStatusFetcher.class);
+
+    private final TaskId taskId; // TODO: Remove after removing references
+    private final StateMachine<TaskStatus> taskStatus; // TODO: Remove after removing references
+
+    private final Consumer<Throwable> onFail;
+    private final Codec<List<TaskStatus>> taskListStatusCodec;
+    private final Duration refreshMaxWait;
+    private final Executor executor;
+    private final HttpClient httpClient;
+    private final RequestErrorTracker errorTracker;
+    private final RemoteTaskStats stats;
+    private final boolean isBinaryTransportEnabled;
+
+    private final AtomicLong currentRequestStartNanos = new AtomicLong();
+
+    @GuardedBy("this")
+    private boolean running;
+
+    @GuardedBy("this")
+    private ListenableFuture<BaseResponse<TaskStatus>> future;
+
+    public WorkerTaskStatusFetcher() {
+        this(null, null, null, null, null, null, null, null, null, null, false);
+    }
+
+    // TODO: We don't need all the parameters here and the logic should be in CBTSF
+    public WorkerTaskStatusFetcher(
+            Consumer<Throwable> onFail,
+            TaskId taskId,
+            TaskStatus initialTaskStatus,
+            Duration refreshMaxWait,
+            Codec<List<TaskStatus>> taskListStatusCodec,
+            Executor executor,
+            HttpClient httpClient,
+            Duration maxErrorDuration,
+            ScheduledExecutorService errorScheduledExecutor,
+            RemoteTaskStats stats,
+            boolean isBinaryTransportEnabled)
+    {
+        idTaskMap = new ConcurrentHashMap<>();
+
+        //requireNonNull(initialTaskStatus, "initialTaskStatus is null");
+
+        this.taskId = requireNonNull(taskId, "taskId is null");
+        this.taskStatus = new StateMachine<>("task-" + taskId, executor, initialTaskStatus);
+        this.onFail = requireNonNull(onFail, "onFail is null");
+
+        this.refreshMaxWait = requireNonNull(refreshMaxWait, "refreshMaxWait is null");
+        this.taskListStatusCodec = requireNonNull(taskListStatusCodec, "taskStatusCodec is null");
+
+        this.executor = requireNonNull(executor, "executor is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        this.errorTracker = new RequestErrorTracker(taskId, initialTaskStatus.getSelf(), maxErrorDuration, errorScheduledExecutor, "getting task status");
+        this.stats = requireNonNull(stats, "stats is null");
+        this.isBinaryTransportEnabled = isBinaryTransportEnabled;
+    }
+
+    public void addTask(TaskId taskId, ContinuousBatchTaskStatusFetcher.Task task) {
```

Comment:
task already have a taskId, why do we need to pass it again here? We could just use it from the task right?